specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2020-04-01/cosmos-db.json

```diff
@@ -1683,6 +1683,86 @@
         }
       }
     },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DocumentDB/databaseAccounts/{accountName}/sqlDatabases/{databaseName}/throughputSettings/default/migrateToAutoscale": {
+      "post": {
```

Comment:
>post [](start = 7, length = 4)

If you model the migration action as POST, it will not be possible to migrate to autoscale declaratively via an ARM template. In addition, it will not be possible to enforce autoscale using Azure Policy (requiring auto scale or auto remediating to enable autoscale seems like a policy that a large org would want).

Why can't we add new properties (or modify existing ones) on the throughput settings resource?

Reply:
We are aware that we would not be able to migrate using ARM templates.

Migration from autoscale to manual and vice-versa is an action, which changes the behavior of the throughputSettings resource. We are not updating the throughputSettings object on in this migration. We have separate PUT operations to update the throughputSettings resource.

Reply:
Usability suffers if you don't support ARM templates. Many large scale customers are using them to declare their infrastructure. To roll this out at scale, they will have write some one off scripts to make a POST or call some cmdlets. We should be making things easier for them and not harder.

In my own services specifically, we have per-geography cosmos db collections across 4 different Azure clouds (and growing). If this worked with a template, I can simply leverage an existing pipeline to safely roll the migration out to all the geographies in all the clouds without having to build something new. And I can make a policy that will block non-autoscaled collections after the migration is done. I can't do any of that when it's modeled as POST.

There are ways we can model this as PUT on throughputSettings that play nice with the rest of the ecosystem. We should explore them...

Reply:
@majastrz, the reason migration is modeled as POST action today is because the system currently does not support user switching modes with the desired configuration in a single API call. Instead, migration action is a parameterless operation where system performs the mode switch and also deduces the new configuration values. After this completes, then user has to make a separate API call to update the configuration in the new mode.

Here is an example. Suppose the resource is in "manual" mode right now, and user wants to switch to "autoscale" mode with a configuration of 50000 maxThroughput. The migration operation today only supports switching the resource to "autoscale" mode, and then assign it a system determined maxThroughput value, which could be 30000. User then has to make a separate API call to update the throughput resource to change maxThroughput to 50000.

Because of this limitation, it would be misleading and confusing to model this as a resource update via PUT, since the API will not be able to honor user's request to change maxThroughput together with the mode switch.

We understand this limitation prevents user from using ARM template for automation. We are in discussion with the team owning the backend system to support seamless mode switches with configuration updates, so that migration can be modeled as a resource update better.